### PR TITLE
devmapper/Remove(): use Rmdir, ignore errors

### PR DESCRIPTION
**Pull request by kolyshkin**
1. Replace EnsureRemoveAll() with Rmdir(), as here we are removing
   the container's mount point, which is already properly unmounted
   and is therefore an empty directory.

2. Ignore the Rmdir() error (but log it unless it's ENOENT). This
   is a mount point, currently unmounted (i.e. an empty directory),
   and an older kernel can return EBUSY if e.g. the mount was
   leaked to other mount namespaces.

### CODE REVIEW

File: daemon/graphdriver/devmapper/driver.go [hunk 1]
```diff
@@ -17,8 +17,9 @@ import (
 	"github.com/docker/docker/pkg/idtools"
 	"github.com/docker/docker/pkg/locker"
 	"github.com/docker/docker/pkg/mount"
-	"github.com/docker/docker/pkg/system"
 	units "github.com/docker/go-units"
+
+	"golang.org/x/sys/unix"
```
Comment (thaJeztah) [inline daemon/graphdriver/devmapper/driver.go]: nit: this (and logrus) should be in the same group;

```patch
diff --git a/daemon/graphdriver/devmapper/driver.go b/daemon/graphdriver/devmapper/driver.go
index e9d9e85aa..83a57f328 100644
--- a/daemon/graphdriver/devmapper/driver.go
+++ b/daemon/graphdriver/devmapper/driver.go
@@ -9,8 +9,6 @@ import (
        "path"
        "strconv"
 
-       "github.com/sirupsen/logrus"
-
        "github.com/docker/docker/daemon/graphdriver"
        "github.com/docker/docker/pkg/containerfs"
        "github.com/docker/docker/pkg/devicemapper"
@@ -18,7 +16,7 @@ import (
        "github.com/docker/docker/pkg/locker"
        "github.com/docker/docker/pkg/mount"
        units "github.com/docker/go-units"
-
+       "github.com/sirupsen/logrus"
        "golang.org/x/sys/unix"
 )
```

File: daemon/graphdriver/devmapper/driver.go [hunk 2]
```diff
@@ -160,7 +161,21 @@ func (d *Driver) Remove(id string) error {
 	if err := d.DeviceSet.DeleteDevice(id, false); err != nil {
 		return fmt.Errorf("failed to remove device %s: %v", id, err)
 	}
-	return system.EnsureRemoveAll(path.Join(d.home, "mnt", id))
+
+	// Most probably the mount point is already removed on Put()
+	// (see DeviceSet.UnmountDevice()), but just in case it was not
+	// let's try to remove it here as well, ignoring errors as with
```
Comment (thaJeztah) [inline daemon/graphdriver/devmapper/driver.go]: nit: s/as with an older kernel can/as an older kernel can/

File: daemon/graphdriver/devmapper/driver.go [hunk 3]
```diff
@@ -160,7 +161,21 @@ func (d *Driver) Remove(id string) error {
...
+	// an older kernel can return EBUSY if e.g. the mount was leaked
+	// to other mount namespaces. A failure to remove the container's
+	// mount point is not important and should not be treated
+	// as a failure to remove the container.
+	mp := path.Join(d.home, "mnt", id)
+	err := unix.Rmdir(mp)
+	if err != nil && !os.IsNotExist(err) {
+		logrus.Warnf("Can't rmdir %s: %s", mp, err)
```
Comment (thaJeztah) [inline daemon/graphdriver/devmapper/driver.go]: perhaps

```go
logrus.Warnf("unable to remove mountpoint %q: %s", mp, err)
```

("unable" or "failed" - used "unable" to make it look less as an important error)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Left some nits, but "SGTM". Would like to have @cpuguy83 to have a look at least :smile:

File: daemon/graphdriver/devmapper/driver.go (hunk 3 above)
Comment (cpuguy83) [reply daemon/graphdriver/devmapper/driver.go]: `WithField("module", "graphdriver/devicemapper")`

**Review by cpuguy83 — APPROVED** [review]
LGTM other than the log message.

**Comment by kolyshkin** [comment]
> WithField("module", "graphdriver/devicemapper")

devmapper uses `logrus.WithField("storage-driver", "devicemapper")` in some other messages.

aufs uses
```
logrus.WithFields(logrus.Fields{
                "module": "graphdriver",
                "driver": "aufs",
        }) 
```

I am in flux about which of the three to use. I guess I'll use the one already used by devmapper -- let me know if you want me to change it.

**Comment by kolyshkin** [comment]
patch updated

**Review by thaJeztah — COMMENTED** [review]
Log message looks ok to me.

Is there a way to test for the situation we’re dealing with?

**Comment by thaJeztah** [comment]
Should we have a tracking issue for standardizing those log properties among storage drivers, or not worth the effort? (could be a “beginners” issue for someone to work on)

**Comment by kolyshkin** [comment]
> Should we have a tracking issue for standardizing those log properties among
> storage drivers, or not worth the effort? (could be a “beginners” issue for someone to work on)

Makes total sense to me

File: daemon/graphdriver/devmapper/driver.go [hunk 4]
```diff
@@ -9,16 +9,16 @@ import (
...
-
 	"github.com/docker/docker/daemon/graphdriver"
 	"github.com/docker/docker/pkg/containerfs"
 	"github.com/docker/docker/pkg/devicemapper"
 	"github.com/docker/docker/pkg/idtools"
 	"github.com/docker/docker/pkg/locker"
 	"github.com/docker/docker/pkg/mount"
-	"github.com/docker/docker/pkg/system"
 	units "github.com/docker/go-units"
+
```
[the remark this answers is not in the document]
Comment (kolyshkin) [reply daemon/graphdriver/devmapper/driver.go]: In this case I wanted to separate "our" packages (ie github.com/docker/docker/*) from third-party ones. But maybe you're right, it makes little sense if the list is sorted, we clearly see docker/docker from others.

Blank line removed.

**Comment by kolyshkin** [comment]
> Is there a way to test for the situation we’re dealing with?

It's a combination of factors that should not happen when using a recent moby/moby, although I guess if you would use an older kernel (like RHEL/CentOS 7.3) and "organize" a mount point leak, you might end up seeing this. It seems way too complicated to setup for me.

**Comment by thaJeztah** [comment]
Opened https://github.com/moby/moby/issues/36478 for tracking the "standardisation" of log-messages

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**